### PR TITLE
Enforced WebSocket-only Socket.IO with an allowRequest configuration

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
- Enforced WebSocket-only Socket.IO with an allowRequest configuration
- Added io.allowedOrigins configuration support and wired it into Helm charts and documentation
- Extended backend tests to cover allowed, disallowed, and default-origin scenarios

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```breaking operator
The backend now requires an `io.AllowedOrigin` configuration to protect the Socket.IO server from cross-site requests

* To allow all origins, you may set the value to `*` (not recommended)
* When deploying with the **Gardener operator**, this setting is automatically configured based on the public ingress hosts
* For manual deployments, you must provide the correct values in the Helm chart configuration
```

### CODE REVIEW

File: backend/lib/io/index.js [hunk 1]
```diff
@@ -9,11 +9,40 @@ import logger from '../logger/index.js'
...
+import config from '../config/index.js'
 
 function init (httpServer, cache) {
+  const allowedOrigins = config.websocketAllowedOrigins
+  if (!allowedOrigins?.length) {
+    logger.warn('websocketAllowedOrigins configuration is required')
+  }
+  allowedOrigins.forEach(allowedOrigin => {
+    logger.info('WebSocket allowed origin: %s', allowedOrigin)
+  })
```
Comment (petersutter) [inline backend/lib/io/index.js]:
```suggestion
  if (!allowedOrigins?.length) {
    logger.warn('websocketAllowedOrigins configuration is required')
  } else {
    logger.info('WebSocket allowed origins: %s', allowedOrigins.join(', '))
  }
```

File: backend/lib/io/index.js [hunk 2]
```diff
@@ -9,11 +9,40 @@ import logger from '../logger/index.js'
...
+
+  const allowAll = allowedOrigins.includes('*')
+  if (allowAll) {
+    logger.warn('WebSocket allowing all origins (*) - this seeting is not recommended for production environments')
```
Comment (petersutter) [inline backend/lib/io/index.js]: in case we want to have different messages depending on the isProd config. Feel free to take over or discard this suggestion

```suggestion
    const msg = config.isProd
      ? 'WebSocket allowing all origins (*) — this is unsafe in production. Restrict allowedOrigins.'
      : 'WebSocket allowing all origins (*) — OK for local/dev, but do not use in production.'
    logger.warn(msg)
```

File: charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js [hunk 3]
```diff
@@ -889,5 +889,20 @@ describe('gardener-dashboard', function () {
         expect(pick(config, ['frontend.experimental'])).toMatchSnapshot()
       })
     })
+
+    describe('socket io', function () {
+      it('should render the template with allowed origins', async function () {
```
Comment (petersutter) [inline charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js]: add test for default * case

File: charts/gardener-dashboard/values.yaml [hunk 4]
```diff
@@ -84,6 +84,12 @@ global:
     # Example: nodeOptions: [--optimize-for-size, --max-old-space-size=920, --gc-interval=100]
     nodeOptions: [--max-old-space-size=920]
 
+    io:
+      # list of allowed origins for the socket.io events endpoint
+      # use ["*"] to allow all origins
```
Comment (petersutter) [inline charts/gardener-dashboard/values.yaml]: mention that * is not recommended for production use cases

File: backend/lib/io/index.js [hunk 5]
```diff
@@ -9,11 +9,40 @@ import logger from '../logger/index.js'
 import helper from './helper.js'
 import dispatcher from './dispatcher.js'
 import _ from 'lodash-es'
+import config from '../config/index.js'
 
 function init (httpServer, cache) {
+  const allowedOrigins = config.websocketAllowedOrigins
+  if (!allowedOrigins?.length) {
+    logger.warn('websocketAllowedOrigins configuration is required')
+  }
```
Comment (petersutter) [inline backend/lib/io/index.js:20]:
```suggestion
```

rm, we don't reach this code as websocketAllowedOrigins is a required property

File: backend/__tests__/config.spec.cjs [hunk 6]
```diff
@@ -167,14 +122,44 @@ describe('config', function () {
...
         expect(config.oidc.issuer).toBe('https://kubernetes:32001')
         expect(config.oidc.client_id).toBe('client_id_from_file')
         expect(config.oidc.client_secret).toBe('client_secret_from_file')
         expect(config.gitHub.authentication.appId).toBe(12345)
         expect(config.gitHub.authentication.installationId).toBe(67890)
+        expect(config.websocketAllowedOrigins).toEqual(['*'])
+        expect(config.logLevel).toBe('info')
+      })
+
+      it('should return the config overridden by environment variables', function () {
```
Comment (petersutter) [inline backend/__tests__/config.spec.cjs:137]: add test for empty websocketAllowedOrigins

Comment (grolu) [reply backend/__tests__/config.spec.cjs:137]: I upgraded the array env parsing logic and added a bunch of tests that ensure required config early exits

File: backend/lib/io/index.js (hunk 5 above)
Comment (grolu) [reply backend/lib/io/index.js:20]: You can specify an empty array via configuration. The env paring logic discards empty arrays but it is still possible via config yaml like this

`websocketAllowedOrigins: []`

Instead of adding a logger warn here I suggest to throw an error in case of empty array. This has basically the same effect as the error is catched upstream and logged as warning but it makes the code cleaner in my pov.

File: backend/lib/io/index.js (hunk 2 above)
Comment (grolu) [reply backend/lib/io/index.js]: lgtm

File: backend/__tests__/config.spec.cjs [hunk 7]
```diff
@@ -85,96 +86,129 @@ describe('config', function () {
...
 
-        const filename = '/etc/gardener/3/config.yaml'
-        const config = gardener.loadConfig(filename, { env })
-        expect(gardener.readConfig).toHaveBeenCalledTimes(1)
-        expect(gardener.readConfig.mock.calls[0]).toEqual([filename])
+        // global env
+        expect(config.apiServerUrl).toBe(env.API_SERVER_URL)
         expect(config.sessionSecret).toBe(env.SESSION_SECRET)
-        expect(config.logLevel).toBe('debug')
+        expect(config.websocketAllowedOrigins).toEqual(env.WEBSOCKET_ALLOWED_ORIGINS.split(','))
```
Comment (petersutter) [inline backend/__tests__/config.spec.cjs]: write it out what the expected value is

File: backend/__tests__/config.spec.cjs (hunk 6 above)
Comment (petersutter) [reply backend/__tests__/config.spec.cjs:137]: I'm still missing a test where config.websocketAllowedOrigins = [] or is not set, both of which should result in an error.

File: backend/lib/config/gardener.js [hunk 8]
```diff
@@ -115,15 +115,29 @@ const configMappings = [
...
+    configPath: 'websocketAllowedOrigins',
+    type: 'Array',
+  },
 ]
 
 function parseConfigValue (value, type) {
+  const parseArray = value => {
+    if (typeof value !== 'string' || !value.length > 0) {
+      return undefined
+    }
```
Comment (petersutter) [inline backend/lib/config/gardener.js]: did you mean to write this odd `!value.length > 0`?

I think it's better to be more explicit
```suggestion
    if (value == null || typeof value !== 'string' || value.length === 0) {
      return undefined;
    }
```

File: backend/lib/io/index.js (hunk 5 above)
Comment (petersutter) [reply backend/lib/io/index.js:20]: an empty array should should not be allowed in the config, at least for websocketAllowedOrigins

File: backend/lib/io/index.js [hunk 9]
```diff
@@ -14,16 +14,17 @@ import config from '../config/index.js'
...
+    logger.info('WebSocket allowed origins: %s', allowedOrigins.join(', '))
   }
-  allowedOrigins.forEach(allowedOrigin => {
-    logger.info('WebSocket allowed origin: %s', allowedOrigin)
-  })
+
+  const msg = config.isProd
+    ? 'WebSocket allowing all origins (*) — this is unsafe in production. Restrict allowedOrigins.'
+    : 'WebSocket allowing all origins (*) — OK for local/dev, but do not use in production.'
+  logger.warn(msg)
```
Comment (petersutter) [inline backend/lib/io/index.js]: this should be inside 
```js
if (allowAll) {
  // log warning here
}
```

File: charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js [hunk 10]
```diff
@@ -903,6 +903,19 @@ describe('gardener-dashboard', function () {
         const config = yaml.load(configMap.data['config.yaml'])
         expect(pick(config, ['io'])).toMatchSnapshot()
       })
+
+      it('should render the template with default allowed origin', async function () {
+        const values = {
+          global: {
+            websocketAllowedOrigins: ['*'],
```
Comment (petersutter) [inline charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js]: unset `websocketAllowedOrigins` so that it defaults to * because of https://github.com/gardener/dashboard/blob/cca93ef7ace9305bc28f260c1730061a72796871/charts/gardener-dashboard/charts/runtime/templates/dashboard/configmap.yaml#L40

I want to test the case ` | default (list "*")`

File: backend/__tests__/config.spec.cjs (hunk 6 above)
Comment (grolu) [reply backend/__tests__/config.spec.cjs:137]: I removed the env variables completely from the test.
The empty array case is tested in `io.cors.spec.cjs` *should reject connections if no origins are configure*

File: charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js (hunk 10 above)
Comment (grolu) [reply charts/__tests__/gardener-dashboard/runtime/dashboard/configmap.spec.js]: Changed the test to unset the values data. Now the helm defaulting is tested.
Also fixed the keys and helm structure for websocketAllowedOrigins
